mlir/lib/Dialect/SCF/Transforms/RotateWhileLoop.cpp

```diff
@@ -0,0 +1,82 @@
+//===- RotateWhileLoop.cpp - scf.while loop rotation ----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Rotates `scf.while` loops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+#include "mlir/Dialect/SCF/Transforms/Patterns.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "scf-rotate-while"
+
+namespace mlir {
+#define GEN_PASS_DEF_SCFROTATEWHILELOOPPASS
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+struct RotateWhileLoopPattern : OpRewritePattern<scf::WhileOp> {
+  RotateWhileLoopPattern(bool rotateLoop, MLIRContext *context,
+                         PatternBenefit benefit = 1,
+                         ArrayRef<StringRef> generatedNames = {})
+      : OpRewritePattern<scf::WhileOp>(context, benefit, generatedNames),
+        forceCreateCheck(rotateLoop) {}
+
+  LogicalResult matchAndRewrite(scf::WhileOp whileOp,
+                                PatternRewriter &rewriter) const final {
+    FailureOr<scf::WhileOp> result =
+        scf::wrapWhileLoopInZeroTripCheck(whileOp, rewriter, forceCreateCheck);
+    if (failed(result) || *result == whileOp) {
+      LLVM_DEBUG(whileOp->emitRemark("Failed to rotate loop"));
+      return failure();
+    };
+    return success();
+  }
+
+  bool forceCreateCheck;
+};
+
+struct SCFRotateWhileLoopPass
+    : impl::SCFRotateWhileLoopPassBase<SCFRotateWhileLoopPass> {
+  using Base::Base;
+
+  void runOnOperation() final {
+    Operation *parentOp = getOperation();
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(context);
+    SCFRotateWhileLoopPassOptions options{forceCreateCheck};
+    scf::populateSCFRotateWhileLoopPatterns(patterns, options);
+    // Avoid applying the pattern to a loop more than once.
+    GreedyRewriteConfig config;
+    config.strictMode = GreedyRewriteStrictness::ExistingOps;
+    [[maybe_unused]] LogicalResult success =
+        applyPatternsAndFoldGreedily(parentOp, std::move(patterns), config);
```

Comment:
Why do we need the fixed-point property of the greedy rewriter here?

The TestSCFWrapInZeroTripCheck.cpp didn't.

Reply:
I'll reply to both your comments here:

If the option forcing guard insertion even when the loop is already in a do-while shape is passed, we'd run into an infinite loop unless we have this fixpoint.

The test wasn't using patterns, but a walker on the `scf.while` operations. I wanted to expose this as a pattern. As I'm gonna get rid of the option in the populate patterns API, this won't be an issue there.

Reply:
My comments stand though, I still don't understand why the pass is using the pattern and greedy rewriter in the first place?

for my other comment: `wrapWhileLoopInZeroTripCheck` does not document when it is returning a failure, which means I can't really understand the pattern behavior (when is it returning success/failure), can you document this?

Reply:
I'm just walking on operations for the pass and not accepting the option in the patterns to fix potential infinite recursion issues now. Also documented when the function fails (it does not fail as of now) and added some more comments all around.

Reply:
The goal, if I understood correctly, is to expose this transformation as a pattern (probably to be composable with other patterns), and this pass only needed to test it. We can keep it as test pass I think (see my other comment).

Reply:
> We can keep it as test pass I think (see my other comment).

Yes it should stay a test pass if the motivation is just to expose the transformation as a pattern (please adjust the PR title and description as well)